### ISSUE TITLE
Questions about the Bayar Convolution(i.e. Constrainted Convolution) Layer in Noise Sensitive Branch.

### ISSUE THREAD

**Issue by Codefmeister** [dong03/MVSS-Net#3]
Hi! Sorry to disturb you again. 
I download the pretrain model from your link, and I found that the constraint conv layer only have one param and it's size is [1, 3, 24]. But I thought that it would be [3, 3, 5, 5] according to your figure.
Could you plz figure out where the wrong lied in?
Thanks sincerely.
![image](https://user-images.githubusercontent.com/57146194/127108857-2e56da4c-e0a4-4f03-9fbb-74d9d3e1447f.png)
![image](https://user-images.githubusercontent.com/57146194/127108902-3e9d1bc5-6f66-4d66-a9c3-ecf8658f0fb8.png)

**Comment by dong03** [comment]
1, we transfer input image from bgr to gray, so the input channel is 1.
and 2, according to _**Constrained convolutional neural networks: A new approach towards general purpose image manipulation detection**_ , the center element of constrained-conv (aka, BayarConv) is constrained to be -1, thus only kernel**2-1 parameters is trainable.  
That is how we get [1, 3, 24].
Implement details of these modules will be included in futher released code.

**Comment by chchshshhh** [comment]
I use constrained convolution to get a different picture after my own training from the weight file constrained convolution. May I ask what is the reason for this?
![image](https://user-images.githubusercontent.com/31399261/132607911-b571524d-e9d8-484c-850a-d30e09b4cb05.png)

![image](https://user-images.githubusercontent.com/31399261/132607845-a022e0e5-653b-46d0-b7c7-d67a98d76c96.png)

**Comment by Chenxr1999** [comment]
> I use constrained convolution to get a different picture after my own training from the weight file constrained convolution. May I ask what is the reason for this?

Sorry I didn't get it. Could you explain how you get the two pictures concretely？Are they from the same original image？

**Comment by chchshshhh** [comment]
The weights obtained by my own training and your pre-trained weights are loaded into the model, and the constrained convolution output is obtained after debugging. The obtained images are shown below, which are different. .
![image](https://user-images.githubusercontent.com/31399261/132864595-e1c9793e-e362-4ace-879e-90ab59bc5828.png)
yours：
![image](https://user-images.githubusercontent.com/31399261/132828147-a24a65fb-baa3-423f-9b2a-06499b63a302.png)
mine:
![image](https://user-images.githubusercontent.com/31399261/132863880-ca112e06-6d57-4c0b-b676-dc07a491ec51.png)



> > I use constrained convolution to get a different picture after my own training from the weight file constrained convolution. May I ask what is the reason for this?
> 
> Sorry I didn't get it. Could you explain how you get the two pictures concretely？Are they from the same original image？

**Comment by dong03** [comment]
I can not figure it out either. The constrained cnn is used to extract the noise pattern inconsistencies, the visualization of it is more likely used to show "well there do have sth strange comparing with context, and such clue may be further mined via CNN". 

One possible reason is that, the output of constrained cnn is not restricted between 0 and 1(or 255), and the matplotlib tend to self-adjust color-map according to the distribution of input values and data type. So I suggust that you may use a sigmoid to observe the relative difference, but not the absolute values. (BTW, the color map of constrained cnn's visualization in *CONSTRAINED R-CNN: A GENERAL IMAGE MANIPULATION DETECTION MODEL* is also different from both of us.

It's just my conjecture, hope it helps.

**Comment by chchshshhh** [comment]
Thank you for your answer, I will try it. Actually, I was thinking about whether the method of data enhancement will make the constrained convolution training produce a relatively large difference in the image, if possible, can you release the code of training data enhancement so that I can have a try, finally I really appreciate the patience to answer these questions

> I can not figure it out either. The constrained cnn is used to extract the noise pattern inconsistencies, the visualization of it is more likely used to show "well there do have sth strange comparing with context, and such clue may be further mined via CNN".
> 
> One possible reason is that, the output of constrained cnn is not restricted between 0 and 1(or 255), and the matplotlib tend to self-adjust color-map according to the distribution of input values and data type. So I suggust that you may use a sigmoid to observe the relative difference, but not the absolute values. (BTW, the color map of constrained cnn's visualization in _CONSTRAINED R-CNN: A GENERAL IMAGE MANIPULATION DETECTION MODEL_ is also different from both of us.
> 
> It's just my conjecture, hope it helps.
